os/logm/logm.c

```diff
@@ -88,6 +88,18 @@ int logm_internal(int flag, int indx, int priority, const char *fmt, va_list ap)
 		&& flag == LOGM_NORMAL && !up_interrupt_context()) {
 
 		flags = irqsave();
+#if defined(CONFIG_LOGM_SYNCHRONOUS) && (CONFIG_LOGM_TIMEOUT_MS != 0)
+#if CONFIG_LOGM_TIMEOUT_MS > 0
+		int count = (CONFIG_LOGM_TIMEOUT_MS + 9) / 10;
+#else
+		int count = -1;
+#endif
+		while (LOGM_STATUS(LOGM_BUFFER_OVERFLOW) && count--) {
+			irqrestore(flags);
+			usleep(10000);
```

Comment:
We can't add any code here (which can spit log by itself) coz, it would end up in recursive logs.

Reply:
Is this comment related to previous one?

What do you mean by 'We can\'t add any code here'?

Care to explain what you are trying to achieve/prevent?

This is a development feature, aimed at not dropping logs when the log spew is too fast for the buffer to compensate, or the console to push out. It may slow logging tread[s], should developer decide so (by enabling this feature).

Are you trying to say that not giving developers any choice and dropping logs (so they do not show at all on the console) when buffer get overflowed is a better solution?

Reply:
logm_internal (...)
{
        usleep(...) -> there are logs inside this call which could be enabled
 }

what would happen when logs inside usleep are enabled through logm?

logm_internal -> usleep -> logm_internal(coz, log in usleep is directed to logm) -> usleep()  --> ends in recursive calls for ever. Hope you got it.

Reply:
For further more, try enabling a log(directed through logm) inside usleep and see what happens.

Reply:
AFAIK the debugs done in usleep callgraph (I guess it is common to scheduler code) only use DEBUGASSERT() macro, which redirects to lldbg, or lldbg directly.
lldbg may, under some circumstances (configuration) forward to logm.

In short, usleep calls logm only if the board crashes.

IIUC right now (without this patch) it means when the board crashes, we might end up with some logs missing, which would be unfortunate. If there is a work around this, I am not aware of it.

This recursion problem could be solved by:
1. passing the right priority to logm (it is LOGM_ERR now, perhaps failing assert should pass higher priority, like LOGM_CRT or LOGM_EMR)
2. In logm_internal handle critical log messages bypassing ones queued in the buffer (they won't print anyway, since up_assert stops the scheduler, disables interrupts and loops endlessly after it has pushed all the dumpstate logs).

@jeongarmy what do you think?

Reply:
He is saying normal debug messages like sdbg in functions called by sleep, I will check it more.
As for assert like DEBUG(), Already there is a functionality similar your suggest 1 in logm.
lldbg, llvdbg are defined with LOGM_LOWPUT flag. (others have LOGM_NORMAL flag).
when LOGM_LOWPUT flag, logm flush buffer (for order) and push the messages through low output directly.
So assert message can be shown.